File: reference/src/compliance/checker.ts
/**
 * Compliance coverage checker.
 * Evaluates which regulatory controls are covered by enabled AI-SDLC controls.
 */

import type { RegulatoryFramework, ControlMapping } from './mappings.js';
import { AI_SDLC_CONTROLS, getMappingsForFramework, REGULATORY_FRAMEWORKS } from './mappings.js';

export interface ComplianceCoverageReport {
  framework: RegulatoryFramework;
  totalControls: number;
  coveredControls: number;
  gaps: ControlMapping[];
  coveragePercent: number;
}

/**
 * Check compliance coverage for a specific regulatory framework.
 *
 * @param enabledControls - Set of AI-SDLC control IDs currently enabled
 * @param framework - The regulatory framework to check against
 */
export function checkCompliance(
  enabledControls: ReadonlySet<string>,
  framework: RegulatoryFramework,
): ComplianceCoverageReport {
  const mappings = getMappingsForFramework(framework);
  const gaps: ControlMapping[] = [];

  for (const mapping of mappings) {
    if (!enabledControls.has(mapping.controlId)) {
      gaps.push(mapping);
    }
  }

  const totalControls = mappings.length;
  const coveredControls = totalControls - gaps.length;
  const coveragePercent = totalControls === 0 ? 100 : (coveredControls / totalControls) * 100;

  return {
    framework,
    totalControls,
    coveredControls,
    gaps,
    coveragePercent,
  };
}

/**
 * Check compliance coverage across all regulatory frameworks.
 *
 * @param enabledControls - Set of AI-SDLC control IDs currently enabled
 */
export function checkAllFrameworks(
  enabledControls: ReadonlySet<string>,
): ComplianceCoverageReport[] {
  return REGULATORY_FRAMEWORKS.map((fw) => checkCompliance(enabledControls, fw));
}

/**
 * Get all available control IDs.
 * Useful for enabling all controls (full coverage).
 */
export function getAllControlIds(): Set<string> {
  return new Set(AI_SDLC_CONTROLS.map((c) => c.id));
}

File: sdk-python/src/ai_sdlc/compliance/checker.py
"""Compliance coverage checker."""

from __future__ import annotations

from dataclasses import dataclass, field

from .mappings import (
    AI_SDLC_CONTROLS,
    REGULATORY_FRAMEWORKS,
    ControlMapping,
    RegulatoryFramework,
    get_mappings_for_framework,
)


@dataclass
class ComplianceCoverageReport:
    framework: RegulatoryFramework
    total_controls: int
    covered_controls: int
    gaps: list[ControlMapping] = field(default_factory=list)
    coverage_percent: float = 0.0


def check_compliance(
    enabled_controls: set[str],
    framework: RegulatoryFramework,
) -> ComplianceCoverageReport:
    """Check compliance coverage for a specific regulatory framework."""
    mappings = get_mappings_for_framework(framework)
    gaps: list[ControlMapping] = []

    for mapping in mappings:
        if mapping.control_id not in enabled_controls:
            gaps.append(mapping)

    total = len(mappings)
    covered = total - len(gaps)
    pct = 100.0 if total == 0 else (covered / total) * 100

    return ComplianceCoverageReport(
        framework=framework,
        total_controls=total,
        covered_controls=covered,
        gaps=gaps,
        coverage_percent=pct,
    )


def check_all_frameworks(
    enabled_controls: set[str],
) -> list[ComplianceCoverageReport]:
    """Check compliance coverage across all regulatory frameworks."""
    return [check_compliance(enabled_controls, fw) for fw in REGULATORY_FRAMEWORKS]


def get_all_control_ids() -> set[str]:
    """Get all available control IDs."""
    return {c.id for c in AI_SDLC_CONTROLS}
